### ISSUE TITLE
Some issues I encountered with hooks

### ISSUE THREAD

**Issue by max4805** [EverestAPI/Everest#66]
First of all, sorry if this is not the right place to report those.

I'm reporting here the issues I encountered with hooks while developing Extended Variants:

### ~~1/ Hooking `Player.ExplodeLaunch` makes the game explode~~ _(fixed)_

For some reason, hooking Player.ExplodeLaunch (even with a hook that does nothing) crashes the game (with `System.AccessViolationException` in the system's event logs on the XNA branch). This happens with both `On.Celeste` hooks and `IL.Celeste` ones.

Here is an example mod doing just that: [ExplodeLaunch.zip](https://github.com/EverestAPI/Everest/files/3882348/ExplodeLaunch.zip). To replicate, go to room 09 in chapter 6 and hit a bumper.

I didn't encounter this issue with any other method I hooked... and I hooked a lot of them. :thinking: Might just be something special with this method or something I'm doing wrong.

### ~~2/ Unhooking IL hooks does not always work~~ _(fixed)_

More specifically, it seems that if two IL hooks are present on a method, unhooking one of them does not work (as in, the hook is still effective). But if there is only one hook, unhooking it works fine.

Here is a mod to replicate: [ILHook.zip](https://github.com/EverestAPI/Everest/files/3882349/ILHook.zip). This hooks Level.Render to enable debug pathfinder rendering and has an option to change the color grade of the level to Core (hot).
* Go to Forsaken City
* Turn on "Level Render Hook Enabled": level turns red
* Turn it off: the level is still red
* Toggling "Nothing Hook Enabled" at this point makes the colors go back to normal. This is the workaround I used in Extended Variants.

### ~~3/ Cannot IL-hook `Player.orig_Update()`~~ _(fixed)_

Trying to IL-hook Player.orig_Update() gives the following error ([full error log](https://github.com/EverestAPI/Everest/files/3900544/error_log.txt)):

```
System.ArgumentException: The type or method has 1 generic parameter(s), but 2 generic argument(s) were provided. A generic argument must be provided for each generic parameter.
```

I didn't look into this in very deep detail, but from what I can see, the call to `Entity.CollideCheck<Solid, NegaBlock>(Vector2)` is the one causing trouble. `Entity.CollideCheck<T>(Vector2)` also exists, so maybe this one gets picked by accident?

**Comment by max4805** [comment]
I can't reproduce the second one (IL hooks not being unhooked) in latest Everest, thanks!

**Comment by max4805** [comment]
Added another issue I just encountered in the list. 😕

**Comment by max4805** [comment]
I looked into the ExplodeLaunch issue into more detail... I'm very unlikely to be able to figure this out on my own, but here are some details that may help.

I ran the game on the FNA branch + Everest 1162 with only [the ExplodeLaunch test mod](https://github.com/EverestAPI/Everest/files/3882348/ExplodeLaunch.zip) installed, and set a breakpoint with dnSpy in `Bumper.OnPlayer` just before ExplodeLaunch is called, then dashed into a bumper.
- The value passed as the first parameter to ExplodeLaunch is a Vector2 with value (2530.625, -2841.016) (memory = `002A1E45409031C5`).
- The hook instead gets (0, 2530.625) as a value for this vector (memory = `00000000002A1E45`). There seems to be some kind of shift going on in memory:
![image](https://user-images.githubusercontent.com/52103563/70866749-b8c47300-1f6d-11ea-854a-98d22a68719c.png)
- Trying to look into Player with dnSpy gives a bunch of "internal debugger errors".
- Continuing execution makes the game crash in the orig method with a `NullReferenceException`.
